mitodl/mit-open
# Sort by coursenums, filtering by department if applicable

# What are the relevant tickets?
Closes #195

# Description (What does it do?)
- Allows search results to be sorted by primary course number
- If there is a departments filter applied, sort by course numbers for that department
- Modifies `LEARNING_RESOURCE_SORTBY_OPTIONS` to be a dict that is used to provide more user-friendly sort options to the user.  For example, the user can sort by `mitcoursenumber` and that will actually sort by `course.course_numbers.sort_coursenum`
- Adjusts the department filter so that it works by department id instead of name (ex: "CMS-W" instead of "Comparative Media Studies/Writing")

# How can this be tested?
- Temporarily modify `learning_resources_search.constants.SOURCE_EXCLUDED_FIELDS` to be an empty list so you will be able to see the `course.course_numbers.sort_coursenum` values.
- Import a bunch of OCW courses, among them `21H.104J+fall_2010` and a few other courses with the department_id "11" (Urban Studies and Planning), plus some other departments.
 
  ```
  ./manage.py backpopulate_ocw_data --course-num 21h-104j-riots-strikes-and-conspiracies-in-american-history-fall-2010
  ```

- Go to this url to sort by course number, filtered by department:
  http://localhost:8063/api/v1/learning_resources_search/?department=11&sortby=mitcoursenumber
- Verify the results are in ascending order by `course.course_numbers.sort_coursenum` for the respective department ("11").  The course mentioned above, `21H.104J+fall_2010`, has a cross-listed id within that department of "11.015J" so it should appear in the correct position relative to the other courses imported from that department.
- Go to this url to sort by course number, unfiltered by department:
  http://localhost:8063/api/v1/learning_resources_search/?sortby=mitcoursenumber
- Verify the results are sorted by `course.course_numbers.sort_coursenum` where `listing_type=primary`, and not by any cross-listed numbers.
- Check that the schema documentation for this endpoint shows some user-friendly sort options:

<img width="1263" alt="Screenshot 2023-11-08 at 1 29 24 PM" src="https://github.com/mitodl/mit-open/assets/187676/29463848-8816-4f61-87ba-0c80e5ed4013">

I did not try to make the departments parameter schema documentation more user-friendly in this PR, leaving that for a future one because the use of a custom field class (`StringArrayField`) for that parameter makes it seem at first glance that doing so might be somewhat involved.
